**Profitable App Profiles for the App Store and Google Play Markets**

The aim of this project is to find mobile app profiles that are profitable for the App Store and Google Play markets. We're working as data analysts for a company that builds Android and iOS mobile apps, and our job is to enable our team of developers to make data-driven decisions with respect to the kind of apps they build.

At our company, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that our revenue for any given app is mostly influenced by the number of users that use our app. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users

**In the code cell below we:**
* Open the dataset using the `open()`, pass in the name of the dataset and assign the output to a variable named `opened_file`
* From the csv python library we import a reader module which we are going to pass in the `opened_file` and assign the result to a variable named `read_file`
* The `read_file` is converted to a list of list using the `list()` method and the output is assigned to a variable named android_app_data


In [4]:
#Opening and exploring the data

#Open the goole Playstore app data set
opened_file = open('googleplaystore.csv', encoding = 'utf8')

from csv import reader

read_file = reader(opened_file)
android_app_data = list(read_file)
andriod_app_data_header = android_app_data[0] # The header row
android_app_data = android_app_data[1:]# The dataset without header row

#Open the iOS App Store data set
opened_file = open('AppleStore.csv', encoding = 'utf8')
read_file = reader(opened_file)
apple_app_data = list(read_file)
apple_app_data_header = apple_app_data[0]
apple_app_data = apple_app_data[1:]

To simplify the exploration of the datasets, we create a function named `explore_data()` that can be used to repeatedly print rows in a readable way.

In [5]:
def explore_data(dataset, start, end, rows_and_columns = False):
    '''
    dataset: the dataset to be explored, list of list
    start: starting index of a slice from the dataset
    end: ending index of a slice from the dataset
    rows_and_columns: prints the number of rows and columns if True
    '''
    sliced_dataset = dataset[start:end]
    for row in sliced_dataset:
        print(row)
        print() # add blank spaces between the rows
        
    if rows_and_columns:
        print('Number of rows: ', len(dataset))#number of rows for the dataset
        print('Number of columns: ', len(dataset[0]))#number of columns for the dataset

#Android App data
print(andriod_app_data_header)
print()
explore_data(android_app_data, 2, 5, True)



['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']

['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']

Number of rows:  10841
Number of columns:  13


We see that the `Google Playstore` dataset has 10841 rows and 13 columns. At a quick glance we can see that the columns that can be of interest in the dataset are: 'App', 'Category', 'Reviews', 'Install', 'Type', 'Price', and 'Genres'

In [6]:
#Apple Store data
print(apple_app_data_header)
print()
explore_data(apple_app_data, 2, 5, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']

['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']

['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']

Number of rows:  7197
Number of columns:  16


We see that the `AppleStore` dataset has 7197 rows and 16 columns. At a quick glance we can see that the columns that can be of interest in the dataset are: 'id', 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'


For a detailed description of each column you can read the [documentation](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps?datasetId=30069&sortBy=dateRun&tab=collaboration)

**Data Cleaning**

The process of preparing our data for analysis is called data cleaning.We do data cleaning before the analysis; it includes removing or correcting wrong data, removing duplicate data, and modifying the data to fit the purpose of our analysis.

In [7]:
#Check if all data rows are of same length

for row in android_app_data:
    if len(row) != len(andriod_app_data_header):#check if length of row is equal to the correct length
        print("The index of the row is: ", android_app_data.index(row))
        print("The length of the row is: ", len(row))
        print("The length of the header row is: ", len(andriod_app_data_header))
        print(andriod_app_data_header)
        print(row)

The index of the row is:  10472
The length of the row is:  12
The length of the header row is:  13
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


We see that for the `Google Playstore` dataset the row at index 10472 has 12 rows instead of 13 and the 'Genres' column is empty.
We delete this row.

In [8]:
#Delete the row with missing data
del android_app_data[10472]

In [9]:
for row in apple_app_data:
    if len(row) != len(apple_app_data_header):#check if length of row is equal to the correct length
        print("The index of the row is: ", apple_app_data.index(row))
        print("The length of the row is: ", len(row))
        print("The length of the header row is: ", len(apple_app_data_header))
        print(apple_app_data_header)
        print(row)

**Removing Duplate Entries: Part One**
 
On further exploring the Google Play data set we notice some apps have duplicate entries. For instance, Instagram has four duplicate entries:

In [10]:
for app in android_app_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [11]:
def unique_duplicates(dataset):
    duplicate_apps = []
    unique_apps = []
    
    for app in dataset:
        name = app[0]
        if name in unique_apps:#is name already in unique apps list?
            duplicate_apps.append(name)#put the name in duplicates list
        else:
            unique_apps.append(name)#name not in unique apps
    
    return duplicate_apps, unique_apps
    

In [12]:
duplicate_apps, unique_apps = unique_duplicates(android_app_data)
print("Number of duplicate apps: ", len(duplicate_apps))
print()
print("Examples of duplicate apps: ", duplicate_apps[:10])

Number of duplicate apps:  1181

Examples of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


We also check for duplicate apps in the apple store dataset

In [13]:
duplicate_apps, unique_apps = unique_duplicates(apple_app_data)
print("Number of duplicate apps: ", len(duplicate_apps))
print()
print("Examples of duplicate apps: ", duplicate_apps[:200])

Number of duplicate apps:  0

Examples of duplicate apps:  []


On close examination of the printed duplicates `Instagram` app we see observe that the difference is only in the fourth column, the `Reviews` column. The different numbers show that the data was collected at different times.
We van use this iinformation to build a criterion for removing duplicates and using the fact that the higher the reviews, the more recent the data and we will only keep the row with the highest number of reviews and remove the other entries for any given app.

**Remove duplicate entries: Part 2**

In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [14]:
print('Expected length: ', len(android_app_data) - 1181)

Expected length:  9659


We see that after removing duplicate entries we will be left with 9659 rows in the Google Play store apps. We create a dictionary `reviews_max` to store the app and the corresponding maximum `reviews` value. 

In [15]:
reviews_max = {}
for app in android_app_data:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews: #is the key app name already in the dictionary and is the value for the key less than the rating? 
        reviews_max[name] = n_reviews #update reviews for the entry
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Use `print()` to make sure everything went well

In [16]:
print('Expected length: ', len(reviews_max))

Expected length:  9659


Using the `reviews_max` dictionary to remove the duplicate rows.
* We start by initialising two empty lists `android_clean[]` and `already_added[]`
* We loop through the `android_app_data` set and for each iteration do the following:
 * assign the app name to a variable named `name`
 * assign `review` of the current app to `n_reviews`
 * add the current row (`app`) to the `android_clean` list and the app name (`name`) to the `already_added` added list if:
   * the number of reviews of the current app matches the number of reviews of that app as described in the `reviews_max` dictionary; and
   * the name of the app is not already in the `already_added` app list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for `reviews_max[name] == n_reviews`, we'll still end up with duplicate entries for some apps.

In [17]:
android_clean = [] # To store our cleaned data set
already_added = [] # To store app names

for app in android_app_data:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
        

Quickly exploring the new data set, and confirm that the number of rows is 9,659

In [18]:
explore_data(android_clean, 2, 5, True)

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']

['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']

['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']

Number of rows:  9659
Number of columns:  13


As expected we have 9659 rows

**Removing Non-English Apps: Part 1**
  
  Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are designed for an English-speaking audience. However, if we explore the data long enough, we'll find that both datasets have apps with names that suggest they are not designed for an English-speaking audience.
  
  If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets:

In [19]:
print(apple_app_data[813][1])
print(apple_app_data[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ




We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in `ord()` function to find out the corresponding encoding number of each character

In [20]:
def is_english(a_string):
    
    for char in a_string:
        if ord(char) > 127:
            return False
    return True
        

In [21]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


The app seems to work fine but some English sympols like emojis, --mdash, etc, which fall outside the ASCII are not being accepted and this can result in removal of useful apps if we use the function in its current form.

In [22]:
print(ord('™'))
print(ord('😜'))

8482
128540


If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. 

In [23]:
def is_english(a_string):
    non_ascii = 0
    for char in a_string:
        if ord(char) > 127:
            non_ascii += 1
        if non_ascii > 3:
            return False
    return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜😜😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Our filter function is still not perfect, but it should be fairly effective for our analysis.

Below, we use the `is_english()` function to filter out the non-English apps for both data sets.

In [24]:
android_english = []
apple_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)

for app in apple_app_data:
    name = app[1]
    if is_english(name):
        apple_english.append(app)
   

explore_data(android_english, 2, 5, True)
print()
explore_data(apple_english, 2, 5, True)

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']

['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']

['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']

Number of rows:  9614
Number of columns:  13

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']

['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']

['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']

Number of ro

We can see that we now have 9614 android apps and 6183 apple store apps.

**Isolating free apps**

So far we done the following on our dataset:
 * removed inaccurate data
 * removed duplicate app entries
 * removed non-English apps
 
As mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our datasets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis.

In [25]:
# identifying the columns describing the app price
print(andriod_app_data_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [26]:
print(apple_app_data_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


We see that the `price` of the app is at index 7 for android apps and 4 for iOS apps

In [27]:
android_final = []
apple_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)

for app in apple_english:
    price = app[4]
    if price == '0.0':
        apple_final.append(app)

In [28]:
print(len(android_final))

8864


In [29]:
print(len(apple_final))

3222


The two datasets have 8864 free Google Playstore apps and 3222 free Apple Store apps and this should be enough for our analysis.

**Most common Apps by Genre: Part 1**

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1) Build a minimal Android version of the app, and add it to Google Play.

2) If the app has a good response from users, we develop it further.

3) If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for the `Genres` and `Category` for the Android apps and `prime_genre` for the Apple store apps columns in our datasets.


**Most common Apps by Genre: Part 2**

We'll build two functions we can use to analyze the frequency tables:

 * One function to generate frequency tables that show percentages
 * Another function we can use to display the percentages in a descending order

In [51]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Use the the `display_table()` function to display the frequency tables of the columns `Genres` index 9, `Category` index 2, and for the AppleStore apps `prime_genre` index 11

We start by examining the frequency table of the AppleStore `prime_genre` column

**Most Common Apps by Genre: Part 3**

In [52]:
display_table(apple_final, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


For the Apple AppStore dataset we observe the following trends among the free English apps:
* The most common genre is `Games` (58.16%) followed by `Entertainment` (7.88%) and `Photo & Video`(4.97%)
* The least common genres are `Navigation`(0.19%), `Medical`(0.19%), `Catalogs`(0.12%) 
* The general impression is that Apple AppStore is dominated by apps that are designed for fun (games, entertainment, photo & video, social networking, sports, music while there are very few apps designed for practical purposes(education, shopping, utilities, productivity, lifestyle health, etc.)
However, it doesn't mean that fun apps have the most users, because abundance does not necessarily mean demand. 

The `Genres` column of the android apps

In [53]:
display_table(android_final, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The `Category` column of the android apps

In [54]:
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 